xds/internal/xdsclient/e2e_test/misc_watchers_test.go

```diff
@@ -0,0 +1,138 @@
+ * limitations under the License.
+ *
+ */
+
+package e2e_test
+
+import (
+	"context"
+	"testing"
+
+	"google.golang.org/grpc/internal/testutils"
+	"google.golang.org/grpc/internal/testutils/xds/e2e"
+	"google.golang.org/grpc/xds/internal/xdsclient"
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource"
+
+	v3routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+)
+
+// TestWatchCallAnotherWatch covers the case where watch() is called inline by a
+// callback. It makes sure it doesn't cause a deadlock.
+func (s) TestWatchCallAnotherWatch(t *testing.T) {
+	overrideFedEnvVar(t)
+
+	// Start an xDS management server and set the option to allow it to respond
+	// to request which only specify a subset of the configured resources.
+	mgmtServer, nodeID, bootstrapContents, _, cleanup := e2e.SetupManagementServer(t, &e2e.ManagementServerOptions{AllowResourceSubset: true})
+	defer cleanup()
+
+	// Create an xDS client with the above bootstrap contents.
+	client, err := xdsclient.NewWithBootstrapContentsForTesting(bootstrapContents)
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	defer client.Close()
+
+	// Configure the management server to with route configuration resources.
+	resources := e2e.UpdateOptions{
+		NodeID: nodeID,
+		Routes: []*v3routepb.RouteConfiguration{
+			e2e.DefaultRouteConfig(rdsName, ldsName, cdsName),
+			e2e.DefaultRouteConfig(rdsNameNewStyle, ldsNameNewStyle, cdsName),
+		},
+		SkipValidation: true,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// Start a watch for one route configuration resource. From the watch
+	// callback of the first resource, register two more watches (one for the
+	// same resource name, which would be satisfied from the cache, and another
+	// for a different resource name, which would be satisfied from the server).
+	updateCh1 := testutils.NewChannel()
+	updateCh2 := testutils.NewChannel()
+	updateCh3 := testutils.NewChannel()
+	var rdsCancel2, rdsCancel3 func()
+	rdsCancel1 := client.WatchRouteConfig(rdsName, func(u xdsresource.RouteConfigUpdate, err error) {
+		updateCh1.Send(xdsresource.RouteConfigUpdateErrTuple{Update: u, Err: err})
+		// Watch for the same resource name.
+		rdsCancel2 = client.WatchRouteConfig(rdsName, func(u xdsresource.RouteConfigUpdate, err error) {
+			updateCh2.Send(xdsresource.RouteConfigUpdateErrTuple{Update: u, Err: err})
+		})
+		// Watch for a different resource name.
+		rdsCancel3 = client.WatchRouteConfig(rdsNameNewStyle, func(u xdsresource.RouteConfigUpdate, err error) {
+			updateCh3.Send(xdsresource.RouteConfigUpdateErrTuple{Update: u, Err: err})
+		})
+	})
+	defer rdsCancel1()
+	defer func() {
+		if rdsCancel2 != nil {
+			rdsCancel2()
+		}
+		if rdsCancel3 != nil {
+			rdsCancel3()
+		}
```

Comment:
Can this not race with the setting of these variables (in the event of an error)?  I.e. if you make watch #2 on `rdsNameNewStyle` instead (or expect it to match the wrong thing), maybe the race detector will fire because `rdsCancel3` could be set while you're running this defer after the `Fatal`.

Do we even need this defer?  Can we cancel the watch inside the callback instead after doing the `Send`?

Reply:
Good point. Moved these to use `t.Cleanup()` and they look much nicer now.